src/sqlfluff/rules/L013.py

```diff
@@ -42,33 +43,36 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
         elements there are.
 
         """
-        if context.segment.is_type("select_clause_element"):
-            if not any(e.is_type("alias_expression") for e in context.segment.segments):
-                types = {
-                    e.get_type() for e in context.segment.segments if e.name != "star"
-                }
-                unallowed_types = types - {
-                    "whitespace",
-                    "newline",
-                    "column_reference",
-                    "wildcard_expression",
-                }
-                if len(unallowed_types) > 0:
-                    # No fixes, because we don't know what the alias should be,
-                    # the user should document it themselves.
-                    if self.allow_scalar:  # type: ignore
-                        # Check *how many* elements there are in the select
-                        # statement. If this is the only one, then we won't
-                        # report an error.
-                        num_elements = sum(
-                            e.is_type("select_clause_element")
-                            for e in context.parent_stack[-1].segments
-                        )
-                        if num_elements > 1:
-                            return LintResult(anchor=context.segment)
-                        else:
-                            return None
-                    else:
-                        # Just error if we don't care.
+        segment = context.surrogates.segment
+        children = segment.children()
+        if segment.all("select_clause_element") and not children.any(
+            "alias_expression"
+        ):
+            types = set(
+                children.select([lambda s: s.name != "star"]).apply(segpred.get_type)
+            )
+            unallowed_types = types - {
+                "whitespace",
+                "newline",
+                "column_reference",
+                "wildcard_expression",
+            }
+            if unallowed_types:
+                # No fixes, because we don't know what the alias should be,
+                # the user should document it themselves.
+                if self.allow_scalar:  # type: ignore
+                    # Check *how many* elements there are in the select
+                    # statement. If this is the only one, then we won't
+                    # report an error.
+                    num_elements = sum(
+                        e.is_type("select_clause_element")
+                        for e in context.parent_stack[-1].segments
+                    )
```

Comment:
could this be converted to use the new API:
```python
immediate_parent = context.surrogates.parent_stack.last()
num_elements = len(immediate_parent.children("select_clause_element"))
```
(provided `last` is made to work as suggested in previous comment)

Also think `len` should work here as we inherit from tuple which is quite intuitive.

Reply:
Good idea! That _almost_ works. I tested this, and the issue is, the first line returns a `BaseSegment`, not a `Segments` object, then the second line fails because we're working with the wrong type of object.

Perhaps we should provide an alternate version of `last()` and `first()` which returns a `Segments`. Or I could change the existing functions? It seems like both versions are useful in certain situations. 🤔

Reply:
so my suggestion here would be to have `last()` and `first()` return a single length `Segments` instance here for consistency (no match to predicates is just a zero length `Segments` instance).
Same goes for `apply()` it should return a `Segments` instance: `return Segments([fn(s) for s in self])`

You can already get a list of base segments via list(Segments) or iterating over them if someone explicitly wants them.

Reply:
Ok, I changed `first()` and `last()` to return a `Segments` instead, and I updated L013 with your suggested change, which works now. 🎉 

We can't change `apply()` as you suggest, because `apply()` returns the result of calling an arbitrary function on the segments, so this may return strings, numbers, whatever; almost certainly **not** segment objects.

The name `apply()`. comes from a very similar Pandas function: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html

Reply:
ahh right makes sense

It's not exactly like pandas because pandas would still return a `pd.Series` object type. It's just the dtype within the column that gets changed rather than the data structure.
i.e.
```python
import pandas as pd

a = pd.DataFrame({"col1": [0, 1, 2]})  # pd.DataFrame
a["col1"] # pd.Series, with `int` as it's `dtype`
a["col1"].apply(lambda x: str(x))  # pd.Series, with `str` as it's `dtype`, but still a pd.Series
```

But makes sense in our case to not return a `Segments` object here